File: src/lib/siteCopy.ts
import type {
  SiteCopyDto,
  SiteCopyUpsertDto,
} from "./api";

export const DEFAULT_SITE_COPY: SiteCopyUpsertDto = {
  navigation: {
    brand: "Rijul",
    hireMe: "Hire Me",
    links: [
      { label: "Home", href: "/" },
      { label: "Services", href: "#services" },
      { label: "Work", href: "#work" },
      { label: "About", href: "#about" },
      { label: "Contact", href: "#contact" },
    ],
  },
  intro: {
    line1: "I DESIGN.",
    line2: "I BUILD.",
    line3: "I EXPERIMENT.",
    body: "I'm {name}, a UI/UX Designer and Developer focused on creating thoughtful digital experiences and modern web applications.",
  },
  about: {
    number: "01 /",
    label: "About",
    stat1Label: "Years Experience",
    stat2Label: "Projects",
    stat3Label: "Technologies",
    statSuffix: "+",
  },
  skills: {
    number: "02 /",
    label: "Skills",
    heading: "Tools I\nWork With.",
    centerLabel: "RIJUL",
  },
  services: {
    number: "03 /",
    label: "Services",
    heading: "What I\nCan Do.",
  },
  work: {
    number: "04 /",
    label: "Selected Work",
    heading: "Things\nI've Built.",
    viewProjectLabel: "View Project",
    separator: "|",
  },
  experience: {
    number: "05 /",
    label: "Experience",
    heading: "The Journey\nSo Far.",
  },
  education: {
    number: "06 /",
    label: "Education",
    heading: "Where I\nLearned.",
    ofConnector: "of",
    dash: "—",
  },
  personal: {
    label: "Beyond the Screen.",
    heading: "Design ✦ Code ✦ Business",
    marqueeWords: ["Design", "Code", "Business"],
    marqueeSeparator: "✦",
    body: "I enjoy working at the intersection of design, technology and business — creating experiences that aren't only visually engaging, but {highlight}useful and practical{/highlight}.",
  },
  contact: {
    number: "07 /",
    label: "Contact",
    headingLine1: "LET'S",
    headingLine2: "MAKE",
    headingLine3: "SOMETHING.",
    body: "Have an idea, project or collaboration in mind? Let's talk.",
    emailLabel: "Email",
    phoneLabel: "Phone",
    phoneNumber: "+977 9746254793",
    formNameLabel: "Your Name",
    formEmailLabel: "Your Email",
    formMessageLabel: "Your Message",
    namePlaceholder: "John Doe",
    emailPlaceholder: "john@example.com",
    messagePlaceholder: "Tell me about your project...",
    submitLabel: "SEND MESSAGE",
    sendingLabel: "SENDING...",
    successTitle: "Message Sent",
    successBody: "Thanks for reaching out! I'll get back to you as soon as possible.",
    sendAnotherLabel: "SEND ANOTHER",
    errorFallback: "Something went wrong. Please try again.",
  },
  footer: {
    navigationHeading: "Navigation",
    contactHeading: "Contact",
    navLinks: [
      { label: "Home", href: "#home" },
      { label: "Intro", href: "#intro" },
      { label: "About", href: "#about" },
      { label: "Skills", href: "#skills" },
      { label: "Services", href: "#services" },
      { label: "Work", href: "#work" },
      { label: "Experience", href: "#experience" },
      { label: "Education", href: "#education" },
      { label: "Contact", href: "#contact" },
    ],
    builtWith: "Built with curiosity.",
  },
  globalUi: {
    cursorDefault: "VIEW",
    cursorHome: "HOME",
    cursorLetsTalk: "LET'S TALK",
    cursorView: "VIEW",
    cursorDownload: "DOWNLOAD",
    cursorDesign: "DESIGN",
    cursorBuild: "BUILD",
    cursorCreate: "CREATE",
    cursorExplore: "EXPLORE",
    cursorOpen: "OPEN",
    cursorBolt: "⚡️",
    heroImageAlt: "Rijul Dhakal Portrait",
  },
};

/**
 * Merges the CMS SiteCopy (which always contains full groups) with the
 * central defaults. Empty strings from the CMS pass through; only missing
 * groups fall back to the default values.
 */
export function resolveSiteCopy(copy: SiteCopyDto | null): SiteCopyUpsertDto {
  if (!copy) return DEFAULT_SITE_COPY;
  return {
    navigation: copy.navigation ?? DEFAULT_SITE_COPY.navigation,
    intro: copy.intro ?? DEFAULT_SITE_COPY.intro,
    about: copy.about ?? DEFAULT_SITE_COPY.about,
    skills: copy.skills ?? DEFAULT_SITE_COPY.skills,
    services: copy.services ?? DEFAULT_SITE_COPY.services,
    work: copy.work ?? DEFAULT_SITE_COPY.work,
    experience: copy.experience ?? DEFAULT_SITE_COPY.experience,
    education: copy.education ?? DEFAULT_SITE_COPY.education,
    personal: copy.personal ?? DEFAULT_SITE_COPY.personal,
    contact: copy.contact ?? DEFAULT_SITE_COPY.contact,
    footer: copy.footer ?? DEFAULT_SITE_COPY.footer,
    globalUi: copy.globalUi ?? DEFAULT_SITE_COPY.globalUi,
  };
}

File: backend/src/Portfolio.Domain/Entities/SiteCopy.cs
using Portfolio.Domain.Common;

namespace Portfolio.Domain.Entities;

/// <summary>
/// Single-row store (like <see cref="TypographySetting"/>) holding every
/// user-visible static text string on the public site. Each group maps to a
/// section of the portfolio and is serialized to its own jsonb column.
/// </summary>
public class SiteCopy : AuditableEntity
{
    public SiteCopyNavigation Navigation { get; set; } = new();
    public SiteCopyIntro Intro { get; set; } = new();
    public SiteCopyAbout About { get; set; } = new();
    public SiteCopySkills Skills { get; set; } = new();
    public SiteCopyServices Services { get; set; } = new();
    public SiteCopyWork Work { get; set; } = new();
    public SiteCopyExperience Experience { get; set; } = new();
    public SiteCopyEducation Education { get; set; } = new();
    public SiteCopyPersonal Personal { get; set; } = new();
    public SiteCopyContact Contact { get; set; } = new();
    public SiteCopyFooter Footer { get; set; } = new();
    public SiteCopyGlobalUi GlobalUi { get; set; } = new();
}

public sealed class NavLink
{
    public string Label { get; set; } = string.Empty;
    public string Href { get; set; } = string.Empty;
}

public sealed class SiteCopyNavigation
{
    public string Brand { get; set; } = string.Empty;
    public string HireMe { get; set; } = string.Empty;
    public List<NavLink> Links { get; set; } = new();
}

public sealed class SiteCopyIntro
{
    public string Line1 { get; set; } = string.Empty;
    public string Line2 { get; set; } = string.Empty;
    public string Line3 { get; set; } = string.Empty;

    /// <summary>Supports the <c>{name}</c> token, rendered as a highlighted span.</summary>
    public string Body { get; set; } = string.Empty;
}

public sealed class SiteCopyAbout
{
    public string Number { get; set; } = string.Empty;
    public string Label { get; set; } = string.Empty;
    public string Stat1Label { get; set; } = string.Empty;
    public string Stat2Label { get; set; } = string.Empty;
    public string Stat3Label { get; set; } = string.Empty;
    public string StatSuffix { get; set; } = string.Empty;
}

public sealed class SiteCopySkills
{
    public string Number { get; set; } = string.Empty;
    public string Label { get; set; } = string.Empty;

    /// <summary>Multi-line, separated by <c>\n</c>; alternating lines render secondary.</summary>
    public string Heading { get; set; } = string.Empty;
    public string CenterLabel { get; set; } = string.Empty;
}

public sealed class SiteCopyServices
{
    public string Number { get; set; } = string.Empty;
    public string Label { get; set; } = string.Empty;
    public string Heading { get; set; } = string.Empty;
}

public sealed class SiteCopyWork
{
    public string Number { get; set; } = string.Empty;
    public string Label { get; set; } = string.Empty;
    public string Heading { get; set; } = string.Empty;
    public string ViewProjectLabel { get; set; } = string.Empty;
    public string Separator { get; set; } = string.Empty;
}

public sealed class SiteCopyExperience
{
    public string Number { get; set; } = string.Empty;
    public string Label { get; set; } = string.Empty;
    public string Heading { get; set; } = string.Empty;
}

public sealed class SiteCopyEducation
{
    public string Number { get; set; } = string.Empty;
    public string Label { get; set; } = string.Empty;
    public string Heading { get; set; } = string.Empty;
    public string OfConnector { get; set; } = string.Empty;
    public string Dash { get; set; } = string.Empty;
}

public sealed class SiteCopyPersonal
{
    public string Label { get; set; } = string.Empty;
    public string Heading { get; set; } = string.Empty;
    public List<string> MarqueeWords { get; set; } = new();
    public string MarqueeSeparator { get; set; } = string.Empty;

    /// <summary>Supports the <c>{highlight}...{/highlight}</c> token, rendered bold + underlined.</summary>
    public string Body { get; set; } = string.Empty;
}

public sealed class SiteCopyContact
{
    public string Number { get; set; } = string.Empty;
    public string Label { get; set; } = string.Empty;
    public string HeadingLine1 { get; set; } = string.Empty;
    public string HeadingLine2 { get; set; } = string.Empty;

    /// <summary>Rendered with the electric accent color.</summary>
    public string HeadingLine3 { get; set; } = string.Empty;
    public string Body { get; set; } = string.Empty;
    public string EmailLabel { get; set; } = string.Empty;
    public string PhoneLabel { get; set; } = string.Empty;
    public string PhoneNumber { get; set; } = string.Empty;
    public string FormNameLabel { get; set; } = string.Empty;
    public string FormEmailLabel { get; set; } = string.Empty;
    public string FormMessageLabel { get; set; } = string.Empty;
    public string NamePlaceholder { get; set; } = string.Empty;
    public string EmailPlaceholder { get; set; } = string.Empty;
    public string MessagePlaceholder { get; set; } = string.Empty;
    public string SubmitLabel { get; set; } = string.Empty;
    public string SendingLabel { get; set; } = string.Empty;
    public string SuccessTitle { get; set; } = string.Empty;
    public string SuccessBody { get; set; } = string.Empty;
    public string SendAnotherLabel { get; set; } = string.Empty;
    public string ErrorFallback { get; set; } = string.Empty;
}

public sealed class SiteCopyFooter
{
    public string NavigationHeading { get; set; } = string.Empty;
    public string ContactHeading { get; set; } = string.Empty;
    public List<NavLink> NavLinks { get; set; } = new();
    public string BuiltWith { get; set; } = string.Empty;
}

public sealed class SiteCopyGlobalUi
{
    public string CursorDefault { get; set; } = string.Empty;
    public string CursorHome { get; set; } = string.Empty;
    public string CursorLetsTalk { get; set; } = string.Empty;
    public string CursorView { get; set; } = string.Empty;
    public string CursorDownload { get; set; } = string.Empty;
    public string CursorDesign { get; set; } = string.Empty;
    public string CursorBuild { get; set; } = string.Empty;
    public string CursorCreate { get; set; } = string.Empty;
    public string CursorExplore { get; set; } = string.Empty;
    public string CursorOpen { get; set; } = string.Empty;
    public string CursorBolt { get; set; } = string.Empty;
    public string HeroImageAlt { get; set; } = string.Empty;
}
